cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/Job.java

```diff
@@ -40,27 +43,27 @@
     /**
      * The Job is awaiting fulfillment of its Trigger.
      */
-    PENDING_TRIGGER,
+    PENDING_LAUNCH(Collections.<State>emptySet()),
     /**
      * The Job is awaiting satisfaction of all of its constraints.
      */
-    PENDING_CONSTRAINT,
+    PENDING_CONSTRAINT(EnumSet.of(PENDING_LAUNCH)),
     /**
      * The Job is awaiting launch.
      */
-    PENDING_LAUNCH;
+    PENDING_TRIGGER(EnumSet.of(PENDING_CONSTRAINT));
 
-    private static final Map<State, State> possibleTransitions = new HashMap<>();
-    static {
-      possibleTransitions.put(PENDING_TRIGGER, PENDING_CONSTRAINT);
-      possibleTransitions.put(PENDING_CONSTRAINT, PENDING_LAUNCH);
+    private final Set<State> allowedNextState;
+
+    State(Set<State> allowedNextState) {
+      this.allowedNextState = allowedNextState;
     }
 
     /**
      * Checks that the specified State is a valid transition from the current state.
      */
     void checkTransition(State nextState) {
-      Preconditions.checkArgument(possibleTransitions.get(this) == nextState,
+      Preconditions.checkArgument(allowedNextState.contains(nextState),
```

Comment:
If its IllegalStateException, that implies that the `Job` is in an invalid state.
In this case, though, its the entity acting on the `Job` attempting to transition the Job to a state, via a transition that is not allowed, so its the calling code that is attempting a transition with an illegal/unallowed argument.